### PR TITLE
Circle

### PR DESCRIPTION
**Pull request by Irev-Dev**
---
Remaining tasks
- [x] Draft circle looks funny
- [x] Cursor should be crosshair when circle tool is equipped
- [x] Console error when doing the second click for adding the circle `Expected [..] to be a sketchGroup, but it wasn't.`
- [x] Extra segment handle is appearing when it shouldn't
- [x] Fix sketch mode tear down weirdness
- [x] Disable line tool after rectangle or circle
- [x] Overlays [absX, absY, and radius]
- [x] Check line thickness and handles scale correctly with zoom
- [x] Check exiting out of the sketch after the first click but before the second doesn't do anything weird
- [x] fix error when dragging circle's circumference
- [x] Check three dot menu works for circle, (probably need to dump delete short term until multi-profile sketches)
- [x] Figure what's going on with the cargo tests
- [x] Make sure circle sketches can be extruded (with UI)

---

Tests
- [x] draft circle should look right (snapshot)
- [x] Can create a new sketch and put a circle on it.
- [x] Can edit an existing circle changing its center and radius
- [x] test for the overlay
- [x] point and click extrude should work with circle sketches.
---

### CODE REVIEW

File: src/wasm-lib/kcl/src/std/shapes.rs [hunk 1]
```diff
@@ -21,12 +22,24 @@ pub enum SketchSurfaceOrGroup {
     SketchGroup(Box<SketchGroup>),
 }
 
+/// Data for drawing an angled line that intersects with a given line.
```
Comment (jtran) [inline src/wasm-lib/kcl/src/std/shapes.rs:25]: This needs updating.

**Comment by jtran** [comment]
Doc comments that use `circle()` need to get updated. Maybe `shell.rs` was added in your last rebase?

https://github.com/KittyCAD/modeling-app/blob/db6036f3261b1a01e69c86e84aa5bc6f04ba5dff/src/wasm-lib/kcl/src/std/shell.rs#L118

**Comment by jtran** [comment]
This example is failing.

https://github.com/KittyCAD/modeling-app/blob/db6036f3261b1a01e69c86e84aa5bc6f04ba5dff/src/wasm-lib/kcl/src/std/convert.rs#L52-L58

When extruding a circle, `do_post_extrude()` seems to be expecting a `Path::ToPoint` to extract an edge ID, but it's not finding it with the circle. Not sure what needs to change here.

https://github.com/KittyCAD/modeling-app/blob/db6036f3261b1a01e69c86e84aa5bc6f04ba5dff/src/wasm-lib/kcl/src/std/extrude.rs#L147

**Comment by Irev-Dev** [comment]
That is super helpful @jtran thank you.

I think it's because https://github.com/KittyCAD/modeling-app/pull/3860/files#diff-b007215ec7d21c68d6a4487af28c6a25fa8cf5fc95ae008e07e5680261cab1f1L73-L80

circle used to call inner_arc which used ToPoint, but now doesn't. I'm not 100% sure what I need to update yet but at least makes sense why it's unhappy.

File: src/wasm-lib/kcl/src/std/shapes.rs (hunk 1 above)
Comment (Irev-Dev) [reply src/wasm-lib/kcl/src/std/shapes.rs:25]: Ah thanks.

File: src/wasm-lib/kcl/src/std/shapes.rs [hunk 2]
```diff
@@ -66,28 +78,71 @@ async fn inner_circle(
...
         args.clone(),
     )
     .await?;
 
-    // Call arc.
-    sketch_group = crate::std::sketch::inner_arc(
-        crate::std::sketch::ArcData::AnglesAndRadius {
-            angle_start: 0.0,
-            angle_end: 360.0,
-            radius,
```
Comment (Irev-Dev) [inline src/wasm-lib/kcl/src/std/shapes.rs:83]: inner_arc could not be used because under the hood it's using straight line types "toPoint", that's going to have to change but I was worried about making a possible breaking when I'm just trying to get circle in.

File: src/wasm-lib/kcl/src/std/shapes.rs [hunk 3]
```diff
@@ -66,28 +78,71 @@ async fn inner_circle(
...
+            path_id: new_sketch_group.id,
         },
-        sketch_group,
-        tag,
-        args.clone(),
     )
     .await?;
 
-    // Call close.
-    crate::std::sketch::inner_close(sketch_group, None, args).await
```
Comment (Irev-Dev) [inline src/wasm-lib/kcl/src/std/shapes.rs:92]: Using inner_close was a no go as well, because while we do want send the close command to the engine, calling this adds a new segment to the sketch group which doesn't make sense, circle is just one segment.

File: e2e/playwright/authenticatedAppFixture.ts
Comment (Irev-Dev) [inline e2e/playwright/authenticatedAppFixture.ts:1]: start of a playwright fixture, related to point-click.spect.ts

File: e2e/playwright/point-click.spec.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        tabNames: ['Selection: 1 face', 'Distance: 5'],
+        currentArg: '',
+        currentTab: '',
+        currentArgValue: '',
+      })
+    await app.progressCmdBar()
+
+    await app.expectEditor.toContain(expectString)
+  })
+})
```
Comment (Irev-Dev) [inline e2e/playwright/point-click.spec.ts:56]: So hopefully this reads well, I'm not adhering 100% to some of Joel's suggestions, in part because that would require a lot more setup just to get going, (e.g. his whole app waiting concept)

But what I have done with this:
- No use of the playwright API at all in the test (expect doesn't count)
- No timeouts use in any part of this (including the fixture)
- Dabbled with the serialisation with the cmd bar

The serialisation of the cmd bar I find particularly interesting, because in the past I've basically just grabbed the bare minimum from the HTML to progress the test (check it says extrude somewhere or similar, chick the arrow button), but with the serialisation, it definitely help express what we expect the state of the cmd bar to be as it's progressed through.

**Comment by Irev-Dev** [comment]
@franknoirot I know you were interested in getting these kinds of additions to a point where code doesn't have to be sprinkled everywhere.

I think https://github.com/KittyCAD/modeling-app/pull/3859 should have definitely helped somewhat, at least [segments.ts](https://github.com/KittyCAD/modeling-app/pull/3860/files#diff-298459cc99a3e1e07efb0fca18f9d6ae6f869849b07b44f4c526b98fd5490f72) at lest contains both the segment init and update functions together, but I think there's probably more after the first pass. I'll trying and look at this PR with fresh eyes tomorrow and note things that could be done in the future.

**Comment by franknoirot** [comment]
This is all exciting. I understand getting this stuff centralized is really hard, it'll take passes no worries. [This snapshot update just above my comment](https://github.com/KittyCAD/modeling-app/pull/3860/commits/03f8eeb5421c1610c82b78c161adc01bc1890797) has me confused; I've never seen our LSP and syntax highlighting fail in a snapshot test.

File: e2e/playwright/point-click.spec.ts (hunk 4 above)
Comment (Irev-Dev) [reply e2e/playwright/point-click.spec.ts:56]: Talked a bit with Frank about improvements to the cmd bar serilsation, I may or may not make those changes in this PR.

the tabs for example should be called args and should be a key: value object.
We also want a union type for the three main states of the command bar, picking a command, adding arguments and reviewing.

Comment (Irev-Dev) [reply e2e/playwright/point-click.spec.ts]: Okay I think it's better now
image

File: e2e/playwright/authenticatedAppFixture.ts [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        await arrowButton.click()
+      } else {
+        await this.page
+          .getByRole('button', { name: 'checkmark Submit command' })
+          .click()
+      }
+    } else {
+      await this.page.keyboard.press('Enter')
+    }
+  }
```
Comment (Irev-Dev) [inline e2e/playwright/authenticatedAppFixture.ts:148]: Definitely a controversial idea here with the `Math.random()`.

My reasoning was that if these fixtures become common place, and lets say it clicks the button, than we only test that path. Sure we can add a tests that tests 'Enter' specifically, but figured so long as this doesn't make tests brittle (and doesn't seem to ) than we're always verifying that both methods of progressing the cmd bar work.

**Comment by Irev-Dev** [comment]
> I've never seen our LSP and syntax highlighting fail in a snapshot test

I don't think it failed, it was just slow, and I've seen it before, test probably needs some tweaks to make sure the LSP is online before snapshotting.

File: src/clientSideScene/segments.ts
Comment (franknoirot) [inline src/clientSideScene/segments.ts:1]: C: This is a great demonstration of your refactor's impact, great work

File: src/components/ModelingMachineProvider.tsx [hunk 6]
```diff
@@ -41,7 +41,7 @@ import {
   canSweepSelection,
   handleSelectionBatch,
   isSelectionLastLine,
-  isRangeInbetweenCharacters,
+  isRangeInbetweenCharacters as isRangeBetweenCharacters,
```
Comment (franknoirot) [inline src/components/ModelingMachineProvider.tsx:44]: C: Probably just trying to keep the unrelated changes down here yeah?

File: src/lib/toolbar.ts [hunk 7]
```diff
@@ -370,10 +377,22 @@ export const toolbarConfig: Record<ToolbarModeName, ToolbarMode> = {
...
+              },
+            }),
           icon: 'circle',
-          status: 'unavailable',
+          status: 'available',
           title: 'Center circle',
+          disabled: (state) =>
+            !canRectangleOrCircleTool(state.context) &&
+            !state.matches({ Sketch: 'Circle tool' }),
+          isActive: (state) => state.matches({ Sketch: 'Circle tool' }),
```
Comment (franknoirot) [inline src/lib/toolbar.ts:395]:
```suggestion
          isActive: (state) => state.matches({ Sketch: 'Circle tool' }),
          hotkey: (state) =>
            state.matches({ Sketch: 'Circle tool' }) ? ['Esc', 'C'] : 'C',
```

**Review by franknoirot — APPROVED** [review]
Hell yeah man! Just adding the hotkey is the only thing I'd like to add before you get this in. I think the radius indicator would be good as a fast follow. The tool works great, the tests seem sound, and the KCL update is good.

I didn't quite grok it while reading, will the old syntax for `circle` continue to be backwards compatible? This might prove pretty disruptive for users if not.

File: e2e/playwright/authenticatedAppFixture.ts [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    const u = await getUtils(this.page)
+    await this.page.addInitScript(async (code) => {
+      localStorage.setItem('persistCode', code)
+      ;(window as any).playwrightSkipFilePicker = true
+    }, code)
+
+    await this.page.setViewportSize({ width: 1000, height: 500 })
+
+    await u.waitForAuthSkipAppStart()
+  }
```
[the remark this answers is not in the document]
Comment (Irev-Dev) [reply e2e/playwright/authenticatedAppFixture.ts:49]: I would prefer `initialiseDesktop`, might as well share as much of the utils as we can.

Atm, at the bottom of this file there's
```ts
test.beforeEach(async ({ context, page }, testInfo) => {
  await setup(context, page, testInfo)
})
```
Which maybe will have to go byby, and instead` setup` goes in `initialise`; 'setupElectron` in `initialiseDesktop`.

File: e2e/playwright/authenticatedAppFixture.ts [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+   *
+   * Expects the viewPort to be 1000x500 */
+  clickNoWhere = () => this.page.mouse.click(998, 60)
+
+  // Toolbars
+  expectExtrudeButtonToBeDisabled = async () =>
+    await expect(this.extrudeButton).toBeDisabled()
+  expectExtrudeButtonToBeEnabled = async () =>
+    await expect(this.extrudeButton).not.toBeDisabled()
+  clickExtrudeButton = async () => await this.extrudeButton.click()
```
[the remark this answers is not in the document]
Comment (Irev-Dev) [reply e2e/playwright/authenticatedAppFixture.ts:71]: Do you think abstracting it is a potential problem, or more like the particular API I've gone with could be better.

We could be just make the `this.extrudeButton` public so that it's then just normal playwright?
```ts
await expect(app.extrudeButton).not.toBeDisabled()
await app.extrudeButten.click()
```
That's probably a good call.

File: e2e/playwright/point-click.spec.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
+import { test } from './authenticatedAppFixture'
+
+// test file is for testing point an click code gen functionality that's not sketch mode related
+
+test('verify extruding circle works', async ({ app }) => {
+  const file = await app.getInputFile('test-circle-extrude.kcl')
+  await app.initialise(file)
+  const [clickCircle, moveToCircle] = app.makeMouseHelpers(582, 217)
```
[the remark this answers is not in the document]
Comment (Irev-Dev) [reply e2e/playwright/point-click.spec.ts:8]: tuple pattern is nice since it makes renaming them easy, which I think since we need mouse moves to arbitrary coords, should help us keep the rest of the test readable.

File: e2e/playwright/sketch-tests.spec.ts [hunk 11]
```diff
@@ -344,6 +346,92 @@ test.describe('Sketch tests', () => {
...
+  |> circle({ center: [4.61, -5.01], radius: 8 }, %)`
+      )
+    })
+
+    await page.setViewportSize({ width: 1200, height: 500 })
+
+    await u.waitForAuthSkipAppStart()
+    await expect(
+      page.getByRole('button', { name: 'Start Sketch' })
+    ).not.toBeDisabled()
```
[the remark this answers is not in the document]
Comment (Irev-Dev) [reply e2e/playwright/sketch-tests.spec.ts:366]: Ah okay.

File: e2e/playwright/sketch-tests.spec.ts [hunk 12]
```diff
@@ -344,6 +346,92 @@ test.describe('Sketch tests', () => {
...
+    const dragPX = 40
+
+    await page
+      .getByText('circle({ center: [4.61, -5.01], radius: 8 }, %)')
+      .click()
+    await expect(
+      page.getByRole('button', { name: 'Edit Sketch' })
+    ).toBeVisible()
+    await page.getByRole('button', { name: 'Edit Sketch' }).click()
+    await page.waitForTimeout(400)
```
[the remark this answers is not in the document]
Comment (Irev-Dev) [reply e2e/playwright/sketch-tests.spec.ts:401]: That's a good point, if this ends up being the one exception, that will still be a good point.

I have talked to Mike about it, and it sounds like events that are not responses to websocket requests is a bit architectural problem with how the engine's setup. Ofc it can be done but, but not a priority.

File: src/components/ModelingMachineProvider.tsx (hunk 6 above)
Comment (Irev-Dev) [reply src/components/ModelingMachineProvider.tsx]: nah, I used symbol rename in vscode and didn't check, it sometimes does dodgy stuff like this.

Thanks for picking this up.

**Comment by Irev-Dev** [comment]
`verify extruding circle works` one of the tests I added is failing on windows because of the "no new console errors" setup Ry added. Not sure why only on windows, I'll boot up windows tomorrow and figure out what's going on.

File: e2e/playwright/authenticatedAppFixture.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (franknoirot) [reply e2e/playwright/authenticatedAppFixture.ts:71]: Ah yeah I like that a lot!

[the remark this answers is not in the document]
Comment (Irev-Dev) [reply e2e/playwright/authenticatedAppFixture.ts:71]: sweet, won't do it now, but easy change when we go to add more buttons next.

File: e2e/playwright/authenticatedAppFixture.ts (hunk 5 above)
Comment (jtran) [reply e2e/playwright/authenticatedAppFixture.ts:148]: Can we log which path we take so that when it fails, we know which choice was made and can debug?

Comment (jtran) [reply e2e/playwright/authenticatedAppFixture.ts:148]: Doubt it's worth the effort now, but the other way I've seen randomness used in tests in the Ruby ecosystem is to prevent test order dependence. Tests are run in random order. It outputs a random seed when running. Then if you hit an error, you can re-run the tests with that same seed to reproduce the tests exactly. Really helpful for debugging.

File: src/clientSideScene/sceneEntities.ts [hunk 13]
```diff
@@ -857,6 +894,175 @@ export class SceneEntities {
...
+      sketchPathToNode,
+      forward,
+      up,
+      position: sketchOrigin,
+      maybeModdedAst: _ast,
+      draftExpressionsIndices: { start: 0, end: 0 },
+    })
+
+    sceneInfra.setCallbacks({
+      // eslint-disable-next-line @typescript-eslint/no-misused-promises
```
Comment (jtran) [inline src/clientSideScene/sceneEntities.ts:948]: Can we avoid ignoring the lint in new code? The hope is that we're thoroughly testing new code, whether automated or manual, so if we accidentally change behavior, we'll catch it.

File: src/clientSideScene/sceneEntities.ts [hunk 14]
```diff
@@ -857,6 +894,175 @@ export class SceneEntities {
...
+          0,
+          _ast,
+          orthoFactor,
+          sketchGroup
+        )
+        sgPaths.forEach((seg, index) =>
+          this.updateSegment(seg, index, 0, _ast, orthoFactor, sketchGroup)
+        )
+      },
+      // eslint-disable-next-line @typescript-eslint/no-misused-promises
```
Comment (jtran) [inline src/clientSideScene/sceneEntities.ts:1015]: Same. No ignore, please?

File: src/wasm-lib/kcl/src/std/shapes.rs [hunk 15]
```diff
@@ -66,28 +84,71 @@ async fn inner_circle(
...
-            angle_end: 360.0,
-            radius,
+    let angle_start = Angle::from_degrees(0.0);
+    let angle_end = Angle::from_degrees(360.0);
+
+    if angle_start == angle_end {
+        return Err(KclError::Type(KclErrorDetails {
+            message: "Arc start and end angles must be different".to_string(),
+            source_ranges: vec![args.source_range],
+        }));
```
Comment (jtran) [inline src/wasm-lib/kcl/src/std/shapes.rs:103]: Seems like the condition can never be true. Is this a mistake?

**Review by jtran — APPROVED** [review]
I commented on a couple minor things. But if getting CI to pass was tough, I'd say merge it. Fast follow up PR would be nice though.

File: src/wasm-lib/kcl/src/executor.rs [hunk 16]
```diff
@@ -1438,6 +1438,19 @@ pub enum Path {
         /// arc's direction
         ccw: bool,
     },
+    /// a complete arc
+    Circle {
```
Comment (adamchalmers) [inline src/wasm-lib/kcl/src/executor.rs:1442]: Sorry if this is obvious, but, why is `Circle` a kind of path segment? How can you add more path segments before or after a circle? It doesn't make sense to me to have a circle as part of a sequence of paths. Shouldn't it just be a shortcut for creating a path with 1 segment, a TangentialArc from 0 to 360 degrees?

File: src/wasm-lib/kcl/src/std/shapes.rs [hunk 17]
```diff
@@ -66,28 +84,71 @@ async fn inner_circle(
...
-    // Call arc.
-    sketch_group = crate::std::sketch::inner_arc(
-        crate::std::sketch::ArcData::AnglesAndRadius {
-            angle_start: 0.0,
-            angle_end: 360.0,
-            radius,
+    let angle_start = Angle::from_degrees(0.0);
+    let angle_end = Angle::from_degrees(360.0);
+
+    if angle_start == angle_end {
```
Comment (adamchalmers) [inline src/wasm-lib/kcl/src/std/shapes.rs:99]: You don't need this, clearly angle_start and angle_end are constants and don't need to be checked for equality IMO

File: src/wasm-lib/kcl/src/std/shapes.rs [hunk 18]
```diff
@@ -66,28 +84,71 @@ async fn inner_circle(
...
     )
     .await?;
 
-    // Call arc.
-    sketch_group = crate::std::sketch::inner_arc(
-        crate::std::sketch::ArcData::AnglesAndRadius {
-            angle_start: 0.0,
-            angle_end: 360.0,
-            radius,
+    let angle_start = Angle::from_degrees(0.0);
```
Comment (adamchalmers) [inline src/wasm-lib/kcl/src/std/shapes.rs:96]:
```suggestion
    let angle_start = Angle::zero();
```

File: src/wasm-lib/kcl/src/std/shapes.rs [hunk 19]
```diff
@@ -66,28 +84,71 @@ async fn inner_circle(
...
+    let angle_end = Angle::from_degrees(360.0);
```
Comment (adamchalmers) [inline src/wasm-lib/kcl/src/std/shapes.rs:97]:
```suggestion
    let angle_end = Angle::turn();
```

File: src/wasm-lib/kcl/src/executor.rs (hunk 16 above)
Comment (Irev-Dev) [reply src/wasm-lib/kcl/src/executor.rs:1442]: inner_arc is very messed up because some of the arc structs are messed up, it returns a path segment using `ToPoint`, I didn't want to change how the arc segments work in case that's a breaking change. I know if the other arcs are not working correctly I'm just leaving the dirt under the rug and not doing anything about it, but I plan on taking a second pass where I add arcs to the client side scene in which case the arcs structs in rust will probably need some attention too, in which case I'll likely remove this struct again.

I was talking to @jtran about it too, I'll likely remove all but one of the arcs, I don't think there's a need for tangencialArc and arc for example, I think we got confused between the API for users and the functions available to them vs the data we store under-the-hood, I think we only need a struct for each curve type, so straight lines and arcs at this point, but later splines etc etc.

Do you think it's okay to put this circle struct in here short term?

File: src/wasm-lib/kcl/src/std/shapes.rs (hunk 17 above)
Comment (Irev-Dev) [reply src/wasm-lib/kcl/src/std/shapes.rs:99]: thank you, copy paste error from https://github.com/KittyCAD/modeling-app/blob/dabf256e2bfcede1f36a48b0a73069c46cdc2822/src/wasm-lib/kcl/src/std/sketch.rs#L1583

But yeah doesn't make sense here.

File: src/wasm-lib/kcl/src/std/shapes.rs (hunk 15 above)
Comment (Irev-Dev) [reply src/wasm-lib/kcl/src/std/shapes.rs:103]: thank you, copy paste error from https://github.com/KittyCAD/modeling-app/blob/dabf256e2bfcede1f36a48b0a73069c46cdc2822/src/wasm-lib/kcl/src/std/sketch.rs#L1583

But yeah doesn't make sense here.

File: src/clientSideScene/sceneEntities.ts (hunk 13 above)
Comment (Irev-Dev) [reply src/clientSideScene/sceneEntities.ts:948]: I've cleaned them up.

I left them in because they were added in the exact same pattern as other setup draft segment functions, so refactoring this would be the same as refactoring the earlier stuff.

Sorry I had done so much refactoring between this and #3859 I was just getting to the point where I wanted to get the change in.

File: src/clientSideScene/ClientSideSceneComp.tsx [hunk 20]
```diff
@@ -96,15 +96,29 @@ export const ClientSideScene = ({
...
+    canvas.addEventListener(
+      'mousemove',
+      toSync(sceneInfra.onMouseMove, reportRejection),
+      false
+    )
     canvas.addEventListener('mousedown', sceneInfra.onMouseDown, false)
-    canvas.addEventListener('mouseup', sceneInfra.onMouseUp, false)
+    canvas.addEventListener(
+      'mouseup',
+      toSync(sceneInfra.onMouseUp, reportRejection),
```
Comment (Irev-Dev) [inline src/clientSideScene/ClientSideSceneComp.tsx:107]: @jtran re https://github.com/KittyCAD/modeling-app/pull/3860#discussion_r1768630322

I think this is the appropriate way to make us `toSync` right?
I had to change the return types of these callbacks to be promisy/async, but these went listeners need to be synchronous.

File: src/wasm-lib/kcl/src/executor.rs (hunk 16 above)
Comment (adamchalmers) [reply src/wasm-lib/kcl/src/executor.rs:1442]: Yeah in the interests of getting this merged I'm happy to include it, but it doesn't really make sense. Let's consolidate this when you fix up the arcs though. Could you just make an issue for it, and leave the issue in the source code here as a comment? So we don't lose track of it.

**Review by adamchalmers — APPROVED** [review]

File: src/wasm-lib/kcl/src/executor.rs (hunk 16 above)
Comment (Irev-Dev) [reply src/wasm-lib/kcl/src/executor.rs:1443]: Sounds good 414b11d6e02a58974a48063da3c37bec5ab6ad8a

File: src/clientSideScene/ClientSideSceneComp.tsx (hunk 20 above)
Comment (jtran) [reply src/clientSideScene/ClientSideSceneComp.tsx:107]: Yes, awesome! I didn't expect you to fix all the old ignores too, but this is great. Ideally, we'd get rid of them all eventually.
